From 78c59aded56ac3f9338d12164438d6ff0377f535 Mon Sep 17 00:00:00 2001
From: Ronald Claveau <linux-kernel-dev@aliel.fr>
Date: Wed, 22 Apr 2026 11:52:56 +0200
Subject: [PATCH 91/91] arm64: dts: amlogic: t7: khadas-vim4: Enable PCIE

The VIM4 exposes the T7 PCIe-A port through an M.2 slot driven by
GPIOH_3 as the active-low endpoint reset.

Allocate a 16 MiB restricted-dma-pool below the 3.5 GiB boundary,
aligned to 4 MiB, to satisfy both the BL31 memory-protection SMC and
the kernel swiotlb restricted-DMA framework. Attach the pool to the
controller via memory-region and enable the node.

Signed-off-by: Ronald Claveau <linux-kernel-dev@aliel.fr>
---
 .../dts/amlogic/amlogic-t7-a311d2-khadas-vim4.dts  | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/arch/arm64/boot/dts/amlogic/amlogic-t7-a311d2-khadas-vim4.dts b/arch/arm64/boot/dts/amlogic/amlogic-t7-a311d2-khadas-vim4.dts
index e1d0744eb2df..b799e2613766 100644
--- a/arch/arm64/boot/dts/amlogic/amlogic-t7-a311d2-khadas-vim4.dts
+++ b/arch/arm64/boot/dts/amlogic/amlogic-t7-a311d2-khadas-vim4.dts
@@ -45,6 +45,14 @@ secmon_reserved_bl32: secmon@5300000 {
 			reg = <0x0 0x05300000 0x0 0x2000000>;
 			no-map;
 		};
+
+		/* 16 MiB reserved for PCIe BL31 memory-protection via SMC */
+		pcie_dma_pool: pcie-dma-pool {
+			compatible = "restricted-dma-pool";
+			size = <0x0 0x1000000>;
+			alignment = <0x0 0x400000>;
+			alloc-ranges = <0x0 0x0 0x0 0xe0000000>;
+		};
 	};
 
 	xtal: xtal-clk {
@@ -419,6 +427,12 @@ map {
 	};
 };
 
+&pcieA {
+	status = "okay";
+	memory-region = <&pcie_dma_pool>;
+	reset-gpios = <&gpio GPIOH_3 GPIO_ACTIVE_LOW>;
+};
+
 &pwm_ab {
 	status = "okay";
 	pinctrl-0 = <&pwm_a_pins>;
-- 
2.53.0

